File: python/steambridge/session.py
"""The per-game state machine.

One ``Session`` per connected game process. It mirrors the parts of Steam the
harness can model locally - identity, language, app id, stats and achievements -
so a game can boot, read, write and store as it normally would, and so a
transcript shows what it asked and what it was told.

Two things it deliberately does not do:

* answer policy calls (``SteamAPI_Init``, ``SteamAPI_IsSteamRunning``, the
  interface getters). Those decide whether a game believes Steam is present, so
  they stay "no opinion" unless a scenario scripts them. See handlers.py.
* answer anything about ownership, entitlement or licensing.
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any


@dataclass
class Profile:
    """What one game is told about the world, from a scenario file."""

    name: str = "default"
    app_id: int = 0
    steam_id: int = 76561197960287930
    persona_name: str = "DebugPlayer"
    language: str = "english"
    ui_language: str = "english"
    install_path: str = r"C:\Program Files (x86)\Steam"
    build_id: int = 1
    stats: dict[str, int] = field(default_factory=dict)
    achievements: list[dict] = field(default_factory=list)
    scripted: dict[str, dict] = field(default_factory=dict)

    @staticmethod
    def from_dict(name: str, data: dict) -> "Profile":
        profile = Profile(name=name)
        profile.app_id = int(data.get("app_id", profile.app_id))
        profile.steam_id = int(data.get("steam_id", profile.steam_id))
        profile.persona_name = str(data.get("persona_name", profile.persona_name))
        profile.language = str(data.get("language", profile.language))
        profile.ui_language = str(data.get("ui_language", profile.ui_language))
        profile.install_path = str(data.get("install_path", profile.install_path))
        profile.build_id = int(data.get("build_id", profile.build_id))
        profile.stats = {str(k): int(v) for k, v in dict(data.get("stats", {})).items()}
        profile.achievements = [dict(entry) for entry in data.get("achievements", [])]
        profile.scripted = {str(k): dict(v) for k, v in dict(data.get("scripted", {})).items()}
        return profile

    def achievement_index(self, name: str) -> int | None:
        for index, entry in enumerate(self.achievements):
            if str(entry.get("name", "")) == name:
                return index
        return None


class Session:
    """The live state of one game process."""

    def __init__(self, session_id: str, hello: dict, profile: Profile) -> None:
        self.id = session_id
        self.pid = int(hello.get("pid", 0))
        self.exe = str(hello.get("exe", ""))
        self.arch = str(hello.get("arch", ""))
        self.profile = profile
        self.started = time.time()
        self.calls: list[dict] = []
        self.stats_written: dict[str, int] = {}
        self.achievements_set: list[str] = []

    # -- identity -----------------------------------------------------------

    def describe(self) -> str:
        return f"session {self.id} ({self.exe or 'unknown exe'}, pid {self.pid}, profile '{self.profile.name}')"

    # -- the calls the harness can answer from state ------------------------

    def handle(self, name: str, args: dict) -> tuple[bool, Any, dict | None]:
        """Answer one call, or report that this session has no opinion.

        Returns ``(answered, return_value, out_parameters)``.
        """
        handler = _STATE_HANDLERS.get(name)
        if handler is None:
            return (False, None, None)
        return handler(self, args)

    # -- individual behaviours ---------------------------------------------

    def _get_stat(self, args: dict) -> tuple[bool, Any, dict | None]:
        key = str(args.get("pchName", ""))
        if key not in self.profile.stats:
            # Steam reports failure for a name it does not know, and leaves the
            # caller's variable alone - which is exactly what we do here too.
            return (True, False, None)
        return (True, True, {"pnData": int(self.profile.stats[key])})

    def _set_stat(self, args: dict) -> tuple[bool, Any, dict | None]:
        key = str(args.get("pchName", ""))
        value = int(args.get("nData", 0))
        # Setting an unknown name is accepted and remembered, so a game can
        # invent a stat locally without the scenario having listed it first.
        self.profile.stats[key] = value
        self.stats_written[key] = value
        return (True, True, None)

    def _get_achievement(self, args: dict) -> tuple[bool, Any, dict | None]:
        name = str(args.get("pchName", ""))
        index = self.profile.achievement_index(name)
        if index is None:
            return (True, False, None)
        achieved = bool(self.profile.achievements[index].get("achieved", False))
        return (True, True, {"pbAchieved": achieved})

    def _set_achievement(self, args: dict) -> tuple[bool, Any, dict | None]:
        name = str(args.get("pchName", ""))
        index = self.profile.achievement_index(name)
        if index is None:
            return (True, False, None)
        self.profile.achievements[index]["achieved"] = True
        if name not in self.achievements_set:
            self.achievements_set.append(name)
        return (True, True, None)

    def _achievement_name(self, args: dict) -> tuple[bool, Any, dict | None]:
        index = int(args.get("iAchievement", 0))
        if index < 0 or index >= len(self.profile.achievements):
            return (True, "", None)
        return (True, str(self.profile.achievements[index].get("name", "")), None)


def _const(value: Any):
    def handler(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
        del session, args
        return (True, value, None)

    return handler


def _seconds_since_app_active(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, int(time.time() - session.started), None)


def _server_real_time(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del session, args
    # Steam hands out UTC seconds; a game only ever uses it for clock sanity.
    return (True, int(time.time()), None)


def _profile_app_id(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, session.profile.app_id, None)


def _profile_steam_id(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, session.profile.steam_id, None)


def _profile_persona(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, session.profile.persona_name, None)


def _profile_language(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, session.profile.language, None)


def _profile_ui_language(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, session.profile.ui_language, None)


def _profile_install_path(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, session.profile.install_path, None)


def _profile_build_id(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, session.profile.build_id, None)


def _request_current_stats(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    # In real Steam this only queues the request and the answer arrives as a
    # UserStatsReceived callback. Callback injection is the next feature; for now
    # the stats are simply readable, which is what a game does right afterwards.
    del session
    return (True, True, None)


def _store_stats(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    del session
    return (True, True, None)


def _num_achievements(session: Session, args: dict) -> tuple[bool, Any, dict | None]:
    del args
    return (True, len(session.profile.achievements), None)


# Calls answered from session state. Everything absent here is either scripted by
# the scenario or reported as "no opinion", which makes the stub fall back to the
# value it would use with Steam not running.
_STATE_HANDLERS = {
    "SteamAPI_GetHSteamUser": _const(1),
    "SteamAPI_GetHSteamPipe": _const(1),
    "SteamAPI_GetSteamInstallPath": _profile_install_path,
    "SteamAPI_ISteamUser_GetSteamID": _profile_steam_id,
    "SteamAPI_ISteamFriends_GetPersonaName": _profile_persona,
    "SteamAPI_ISteamUtils_GetAppID": _profile_app_id,
    "SteamAPI_ISteamUtils_GetCurrentGameLanguage": _profile_language,
    "SteamAPI_ISteamUtils_GetSteamUILanguage": _profile_ui_language,
    "SteamAPI_ISteamUtils_GetSecondsSinceAppActive": _seconds_since_app_active,
    "SteamAPI_ISteamUtils_GetServerRealTime": _server_real_time,
    "SteamAPI_ISteamApps_GetAppBuildId": _profile_build_id,
    "SteamAPI_ISteamUserStats_RequestCurrentStats": _request_current_stats,
    "SteamAPI_ISteamUserStats_StoreStats": _store_stats,
}

# These need the argument object, so they are registered separately rather than
# through the _const shortcut.
_ARG_HANDLERS = {
    "SteamAPI_ISteamUserStats_GetStatInt32": Session._get_stat,
    "SteamAPI_ISteamUserStats_SetStatInt32": Session._set_stat,
    "SteamAPI_ISteamUserStats_GetAchievement": Session._get_achievement,
    "SteamAPI_ISteamUserStats_SetAchievement": Session._set_achievement,
    "SteamAPI_ISteamUserStats_GetNumAchievements": _num_achievements,
    "SteamAPI_ISteamUserStats_GetAchievementName": Session._achievement_name,
}

_STATE_HANDLERS.update(_ARG_HANDLERS)

# Exposed so handlers.py can ask "does the session know this one?" without
# reaching into the table itself.
STATE_HANDLED_CALLS = frozenset(_STATE_HANDLERS)

File: src/session.cpp
#include "bridge/session.hpp"

#include <chrono>
#include <string>
#include <utility>

namespace steambridge {
namespace {

// ---------------------------------------------------------------------------
//  Tolerant readers.
// ---------------------------------------------------------------------------
//  A scenario is hand written, so a value can arrive as the "wrong" JSON kind -
//  an app id spelled as a string, say. Python's int()/str() coerced those; these
//  do the same, and fall back to the default instead of raising, so a typo in a
//  scenario cannot take the backend down in the middle of a run.

std::int64_t to_int64(const Json& value, std::int64_t fallback) noexcept {
    switch (value.kind()) {
        case Json::Kind::number: return value.as_int64();
        case Json::Kind::boolean: return value.as_bool() ? 1 : 0;
        case Json::Kind::string: {
            const std::string& text = value.as_string();
            std::size_t index = 0;
            while (index < text.size() && (text[index] == ' ' || text[index] == '\t')) {
                ++index;
            }
            bool negative = false;
            if (index < text.size() && (text[index] == '+' || text[index] == '-')) {
                negative = text[index] == '-';
                ++index;
            }
            const std::size_t digits_begin = index;
            std::int64_t magnitude = 0;
            while (index < text.size() && text[index] >= '0' && text[index] <= '9') {
                if (magnitude > (9223372036854775807LL - (text[index] - '0')) / 10) {
                    return fallback;
                }
                magnitude = magnitude * 10 + (text[index] - '0');
                ++index;
            }
            if (index == digits_begin) {
                return fallback;
            }
            while (index < text.size() && (text[index] == ' ' || text[index] == '\t')) {
                ++index;
            }
            if (index != text.size()) {
                return fallback;
            }
            return negative ? -magnitude : magnitude;
        }
        default: return fallback;
    }
}

std::string to_text(const Json& value, const std::string& fallback) {
    switch (value.kind()) {
        case Json::Kind::string: return value.as_string();
        case Json::Kind::number:
            // Only an integral number has a text form worth reading as a name;
            // anything else is more likely a mistake than a name. The cast says
            // out loud what the comparison used to do implicitly.
            return static_cast<double>(value.as_int64()) == value.as_double()
                       ? std::to_string(value.as_int64())
                       : fallback;
        default: return fallback;
    }
}

std::string string_member(const Json& args, const char* key) {
    const Json* value = args.find(key);
    return value != nullptr ? to_text(*value, std::string()) : std::string();
}

Answer answered(Json ret) {
    Answer answer;
    answer.answered = true;
    answer.ret = std::move(ret);
    return answer;
}

Answer answered_with_out(Json ret, Json out) {
    Answer answer;
    answer.answered = true;
    answer.ret = std::move(ret);
    answer.out = std::move(out);
    return answer;
}

// Every state answer is "state"; the one place that changes it is the dispatcher,
// which relabels a scripted win.
constexpr const char* kStateVia = "state";

Answer from_state(Json ret) {
    Answer answer = answered(std::move(ret));
    answer.via = kStateVia;
    return answer;
}

Answer from_state_out(Json ret, Json out) {
    Answer answer = answered_with_out(std::move(ret), std::move(out));
    answer.via = kStateVia;
    return answer;
}

// --- the handlers ----------------------------------------------------------
// Each answers one call from the session's own state. A handler that does not
// recognise a name or an index reports Steam's failure value and leaves the
// caller's variable alone, which is what the stub's "no out-parameter" rule
// already does for us.

Answer h_const_one(Session&, const Json&) { return from_state(Json::integer(1)); }

Answer h_install_path(Session& session, const Json&) {
    return from_state(Json::string(session.profile().install_path));
}

Answer h_steam_id(Session& session, const Json&) {
    return from_state(Json::integer(static_cast<std::int64_t>(session.profile().steam_id)));
}

Answer h_persona_name(Session& session, const Json&) {
    return from_state(Json::string(session.profile().persona_name));
}

Answer h_app_id(Session& session, const Json&) {
    return from_state(Json::integer(session.profile().app_id));
}

Answer h_language(Session& session, const Json&) {
    return from_state(Json::string(session.profile().language));
}

Answer h_ui_language(Session& session, const Json&) {
    return from_state(Json::string(session.profile().ui_language));
}

Answer h_seconds_since_active(Session& session, const Json&) {
    const auto elapsed = std::chrono::system_clock::now() - session.started();
    const auto seconds = std::chrono::duration_cast<std::chrono::seconds>(elapsed).count();
    return from_state(Json::integer(static_cast<std::int64_t>(seconds)));
}

Answer h_server_real_time(Session&, const Json&) {
    // Steam hands out UTC seconds; a game only ever uses it for clock sanity.
    const auto now = std::chrono::system_clock::now().time_since_epoch();
    const auto seconds = std::chrono::duration_cast<std::chrono::seconds>(now).count();
    return from_state(Json::integer(static_cast<std::int64_t>(seconds)));
}

Answer h_build_id(Session& session, const Json&) {
    return from_state(Json::integer(session.profile().build_id));
}

Answer h_true(Session&, const Json&) { return from_state(Json::boolean(true)); }

Answer h_get_stat(Session& session, const Json& args) {
    const std::int64_t* value = session.profile().find_stat(string_member(args, "pchName"));
    if (value == nullptr) {
        // Steam reports failure for a name it does not know, and leaves the
        // caller's variable alone - which is exactly what we do here too.
        return from_state(Json::boolean(false));
    }
    Json out = Json::object();
    out.set("pData", Json::integer(*value));
    return from_state_out(Json::boolean(true), std::move(out));
}

Answer h_set_stat(Session& session, const Json& args) {
    // Setting an unknown name is accepted and remembered, so a game can invent a
    // stat locally without the scenario having listed it first.
    const std::string key = string_member(args, "pchName");
    const Json* value = args.find("nData");
    const std::int64_t number = value != nullptr ? to_int64(*value, 0) : 0;
    session.profile().set_stat(key, number);
    session.note_stat_written(key, number);
    return from_state(Json::boolean(true));
}

Answer h_get_achievement(Session& session, const Json& args) {
    const int index = session.profile().achievement_index(string_member(args, "pchName"));
    if (index < 0) {
        return from_state(Json::boolean(false));
    }
    Json out = Json::object();
    out.set(
        "pbAchieved",
        Json::boolean(session.profile().achievements[static_cast<std::size_t>(index)].achieved));
    return from_state_out(Json::boolean(true), std::move(out));
}

Answer h_set_achievement(Session& session, const Json& args) {
    const std::string name = string_member(args, "pchName");
    Achievement* achievement = session.profile().find_achievement(name);
    if (achievement == nullptr) {
        return from_state(Json::boolean(false));
    }
    achievement->achieved = true;
    session.note_achievement_set(name);
    return from_state(Json::boolean(true));
}

Answer h_num_achievements(Session& session, const Json&) {
    return from_state(
        Json::integer(static_cast<std::int64_t>(session.profile().achievements.size())));
}

Answer h_achievement_name(Session& session, const Json& args) {
    const Json* requested = args.find("iAchievement");
    const std::int64_t index = requested != nullptr ? to_int64(*requested, 0) : 0;
    if (index < 0 || index >= static_cast<std::int64_t>(session.profile().achievements.size())) {
        return from_state(Json::string(""));
    }
    return from_state(
        Json::string(session.profile().achievements[static_cast<std::size_t>(index)].name));
}

// Calls answered from session state. Everything absent here is either scripted
// by the scenario or reported as "no opinion", which makes the stub fall back to
// the value it would use with Steam not running.
struct HandlerEntry {
    const char* name;
    Answer (*handler)(Session&, const Json&);
};

constexpr HandlerEntry kHandlers[] = {
    {"SteamAPI_GetHSteamUser", &h_const_one},
    {"SteamAPI_GetHSteamPipe", &h_const_one},
    {"SteamAPI_GetSteamInstallPath", &h_install_path},
    {"SteamAPI_ISteamUser_GetSteamID", &h_steam_id},
    {"SteamAPI_ISteamFriends_GetPersonaName", &h_persona_name},
    {"SteamAPI_ISteamUtils_GetAppID", &h_app_id},
    {"SteamAPI_ISteamUtils_GetCurrentGameLanguage", &h_language},
    {"SteamAPI_ISteamUtils_GetSteamUILanguage", &h_ui_language},
    {"SteamAPI_ISteamUtils_GetSecondsSinceAppActive", &h_seconds_since_active},
    {"SteamAPI_ISteamUtils_GetServerRealTime", &h_server_real_time},
    {"SteamAPI_ISteamApps_GetAppBuildId", &h_build_id},
    {"SteamAPI_ISteamUserStats_RequestCurrentStats", &h_true},
    {"SteamAPI_ISteamUserStats_StoreStats", &h_true},
    {"SteamAPI_ISteamUserStats_GetStatInt32", &h_get_stat},
    {"SteamAPI_ISteamUserStats_SetStatInt32", &h_set_stat},
    {"SteamAPI_ISteamUserStats_GetAchievement", &h_get_achievement},
    {"SteamAPI_ISteamUserStats_SetAchievement", &h_set_achievement},
    {"SteamAPI_ISteamUserStats_GetNumAchievements", &h_num_achievements},
    {"SteamAPI_ISteamUserStats_GetAchievementName", &h_achievement_name},
};

}  // namespace

// ---------------------------------------------------------------------------
//  Profile
// ---------------------------------------------------------------------------

Profile Profile::from_json(const std::string& profile_name, const Json& data) {
    Profile profile;
    profile.name = profile_name;
    if (!data.is_object()) {
        return profile;
    }

    if (const Json* value = data.find("app_id")) {
        profile.app_id = to_int64(*value, profile.app_id);
    }
    if (const Json* value = data.find("steam_id")) {
        profile.steam_id = static_cast<std::uint64_t>(
            to_int64(*value, static_cast<std::int64_t>(profile.steam_id)));
    }
    if (const Json* value = data.find("persona_name")) {
        profile.persona_name = to_text(*value, profile.persona_name);
    }
    if (const Json* value = data.find("language")) {
        profile.language = to_text(*value, profile.language);
    }
    if (const Json* value = data.find("ui_language")) {
        profile.ui_language = to_text(*value, profile.ui_language);
    }
    if (const Json* value = data.find("install_path")) {
        profile.install_path = to_text(*value, profile.install_path);
    }
    if (const Json* value = data.find("build_id")) {
        profile.build_id = to_int64(*value, profile.build_id);
    }

    if (const Json* value = data.find("stats"); value != nullptr && value->is_object()) {
        for (const auto& member : value->members()) {
            profile.stats.emplace_back(member.first, to_int64(member.second, 0));
        }
    }

    if (const Json* value = data.find("achievements"); value != nullptr && value->is_array()) {
        for (const Json& entry : value->items()) {
            if (!entry.is_object()) {
                continue;
            }
            Achievement achievement;
            if (const Json* entry_name = entry.find("name")) {
                achievement.name = to_text(*entry_name, std::string());
            }
            if (const Json* achieved = entry.find("achieved")) {
                achievement.achieved = achieved->as_bool();
            }
            profile.achievements.push_back(std::move(achievement));
        }
    }

    if (const Json* value = data.find("scripted"); value != nullptr && value->is_object()) {
        for (const auto& member : value->members()) {
            profile.scripted.emplace_back(member.first, member.second);
        }
    }

    return profile;
}

std::int64_t* Profile::find_stat(const std::string& key) noexcept {
    for (auto& entry : stats) {
        if (entry.first == key) {
            return &entry.second;
        }
    }
    return nullptr;
}

const std::int64_t* Profile::find_stat(const std::string& key) const noexcept {
    for (const auto& entry : stats) {
        if (entry.first == key) {
            return &entry.second;
        }
    }
    return nullptr;
}

void Profile::set_stat(const std::string& key, std::int64_t value) {
    if (std::int64_t* existing = find_stat(key)) {
        *existing = value;
        return;
    }
    stats.emplace_back(key, value);
}

int Profile::achievement_index(const std::string& achievement_name) const noexcept {
    for (std::size_t index = 0; index < achievements.size(); ++index) {
        if (achievements[index].name == achievement_name) {
            return static_cast<int>(index);
        }
    }
    return -1;
}

Achievement* Profile::find_achievement(const std::string& achievement_name) noexcept {
    const int index = achievement_index(achievement_name);
    return index < 0 ? nullptr : &achievements[static_cast<std::size_t>(index)];
}

const Json* Profile::scripted_for(const std::string& call) const noexcept {
    for (const auto& entry : scripted) {
        if (entry.first == call) {
            return &entry.second;
        }
    }
    return nullptr;
}

// ---------------------------------------------------------------------------
//  Session
// ---------------------------------------------------------------------------

Session::Session(std::string id, const Json& hello, Profile profile)
    : _id(std::move(id)), _profile(std::move(profile)), _started(std::chrono::system_clock::now()) {
    if (hello.is_object()) {
        if (const Json* pid = hello.find("pid")) {
            _pid = to_int64(*pid, 0);
        }
        if (const Json* exe = hello.find("exe")) {
            _exe = exe->is_string() ? exe->as_string() : std::string();
        }
        if (const Json* arch = hello.find("arch")) {
            _arch = arch->is_string() ? arch->as_string() : std::string();
        }
    }
}

void Session::note_stat_written(const std::string& key, std::int64_t value) {
    for (auto& entry : _stats_written) {
        if (entry.first == key) {
            entry.second = value;
            return;
        }
    }
    _stats_written.emplace_back(key, value);
}

void Session::note_achievement_set(const std::string& name) {
    for (const std::string& existing : _achievements_set) {
        if (existing == name) {
            return;
        }
    }
    _achievements_set.push_back(name);
}

std::string Session::describe() const {
    return "session " + _id + " (" + (_exe.empty() ? std::string("unknown exe") : _exe) + ", pid " +
           std::to_string(_pid) + ", profile '" + _profile.name + "')";
}

Answer Session::handle(const std::string& name, const Json& args) {
    for (const HandlerEntry& entry : kHandlers) {
        if (name == entry.name) {
            return entry.handler(*this, args);
        }
    }
    return Answer{};
}

std::vector<std::string> state_handled_calls() {
    std::vector<std::string> names;
    names.reserve(sizeof(kHandlers) / sizeof(kHandlers[0]));
    for (const HandlerEntry& entry : kHandlers) {
        names.emplace_back(entry.name);
    }
    return names;
}

}  // namespace steambridge
